kartothek/io/dask/repartition.py

```diff
@@ -0,0 +1,81 @@
+import math
+from typing import Any, Optional
+
+import dask
+import dask.dataframe as dd
+
+import kartothek.io.dask.dataframe as ktk_ddf
+from kartothek.core.factory import DatasetFactory
+from kartothek.serialization import PredicatesType
+
+
+def calculate_repartition_ratio(
+    dataset_factory: DatasetFactory,
+    table_name: str,
+    predicates: Optional[PredicatesType] = None,
+    *,
+    target_nrows_per_partition: int,
+    dataset_metadata_frac: float,
+) -> dd.core.Scalar:
+    """
+    Calculate a `repartition_ratio` according to the dataset input and `target_nrows_per_partition` (the desired
+    number of rows of each dataframe partition).
+
+    Parameters
+    ----------
+    dataset_factory
+        dataset factory
+    table_name
+        Dataset Table name
+    predicates
+        Optional , but predicates to be applied during calculatin repartition ratio
+    target_nrows_per_partition
+        target num of rows per partition
+    dataset_metadata_frac
+        used during collection of dataset metadata
+    Returns
+    -------
+    repartition_ratio
+        calculated repartition_ratio
+    """
+    dataset_metadata = ktk_ddf.collect_dataset_metadata(
+        factory=dataset_factory,
+        predicates=predicates,
+        table_name=table_name,
+        frac=dataset_metadata_frac,
+    )
+    # Once the kartothek issue is resolved, one should use the rowgroup wise
+    # metadata. This would set up future improvements once kartothek is able to
+    # filter RG predicates while collecting statistics (assuming this will ever
+    # be possible this way)
+    # See https://github.com/JDASoftwareGroup/kartothek/pull/352
+    # and https://issues.apache.org/jira/browse/ARROW-10146
+    expected_nrows_per_partition = dataset_metadata[
+        ["partition_label", "number_rows_total"]
+    ].drop_duplicates()
+
+    mean_nrows_per_partition = expected_nrows_per_partition.number_rows_total.mean()
+    repartition_ratio = target_nrows_per_partition / mean_nrows_per_partition
+    return repartition_ratio
+
+
+def repartition_ddf(ddf: dd.DataFrame, repartition_ratio: float) -> Any:
```

Comment:
Not sure why this wasn't set before if logic was the same
```suggestion
def repartition_ddf(ddf: dd.DataFrame, repartition_ratio: float) -> dd.DataFrame:
```

Reply:
we have to return `Any`  . I am returning the `dask.otpmize()` result. 
I added Optimization logic here, to the `repartition_ddf ()` .

Reply:
Is the return value not a `dd.DataFrame`?

Reply:
The return type is the same as the input type. It might be the case that `dask.optimize` does not expose the types correctly but the output is definitely `dd.DataFrame`

Reply:
done.please verify